### PR TITLE
[FLINK-12823][datastream] PartitionTransformation supports DataExchan…

### PR DESCRIPTION
**Pull request by ifndef-SleePy**
…geMode property

## What is the purpose of the change

* Since `StreamTransformation` would also support batch runner in 1.9, the result partition type of `StreamTransformation` should not be hard-coded with `PIPELINED_BOUNDED`
* We need to provide a way for upper level API upon `StreamTransformation` to configure the result partition type of edge

## Brief change log

* Expose a property `DataExchangeMode` of `PartitionTransformation` as an internal API of `StreamTransformation`
* Pass the `DataExchangeMode` to `StreamEdge`
* `StreamingJobGraphGenerator` chooses the appropriate result partition type based on `DataExchangeMode` of `StreamEdge`

## Verifying this change

* Add an unit test of `StreamingJobGraphGeneratorTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by aljoscha** [comment]
I think we should not reuse the existing `DataExchangeMode` because it is a runtime class in the `flink-runtime` package. At some point we want to decouple the `StreamTransformations` and the `DataStream` API from runtime so we should not introduce more dependencies. I would suggest to add a new enum next to `PartitionTransformation` (in the same package) for this purpose, it should probably also be declared `@PublicEvolving`.

**Comment by ifndef-SleePy** [comment]
Hi @aljoscha

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraph.java [hunk 1]
```diff
@@ -428,7 +436,11 @@ private void addEdgeInternal(Integer upStreamVertexID,
 				}
 			}
 
-			StreamEdge edge = new StreamEdge(upstreamNode, downstreamNode, typeNumber, outputNames, partitioner, outputTag);
+			if (shuffleMode == null) {
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraph.java]: Can it happen that `shuffleMode` is `null`? We should add a null check in `PartitionTransform` and mark the field `@Nonnull`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/PartitionTransformation.java [hunk 2]
```diff
@@ -69,6 +88,15 @@ public PartitionTransformation(StreamTransformation<T> input, StreamPartitioner<
 		return partitioner;
 	}
 
+	/**
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/PartitionTransformation.java]: I don't like these redundant descriptions to much, this can be simplified to:
```
/**
 * Returns the {@link ShuffleMode} of this {@link PartitionTransformation}.
 */
```

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/PartitionTransformation.java [hunk 3]
```diff
@@ -79,6 +107,6 @@ public PartitionTransformation(StreamTransformation<T> input, StreamPartitioner<
 
 	@Override
 	public final void setChainingStrategy(ChainingStrategy strategy) {
-		throw new UnsupportedOperationException("Cannot set chaining strategy on Union Transformation.");
+		throw new UnsupportedOperationException("Cannot set chaining strategy on Partition Transformation.");
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/PartitionTransformation.java]: Nice catch! 😅

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGeneratorTest.java [hunk 4]
```diff
@@ -133,8 +136,8 @@ public void testChainStartEndSetting() throws Exception {
 				public Integer map(Integer value) throws Exception {
 					return value;
 				}
-			})
-			.print();
+			}).setParallelism(2)
```
Comment (aljoscha) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGeneratorTest.java]: Is this change needed? It seems to be an unrelated test.

Comment (ifndef-SleePy) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGeneratorTest.java]: It's an unrelated change in a sense.
I found it when adding unit test. It's a potential unstable case since it depends on the `StreamExecutionEnvironment.defaultLocalParallelism` which must be not 1 in this case.
However I just found someone has fix it yesterday. What a coincident! I can remove it from this PR.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraph.java (hunk 1 above)
Comment (ifndef-SleePy) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraph.java]: Yes, it can be null here. Because there might be no `PartitionTransformation` if user do not specify explicitly. And then `partitioner` and `shuffleMode` would be null here.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/PartitionTransformation.java (hunk 2 above)
Comment (ifndef-SleePy) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/PartitionTransformation.java]: Done

**Comment by ifndef-SleePy** [comment]
Hi @aljoscha 
I have rebased master and addressed comments.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraph.java (hunk 1 above)
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraph.java]: Ok, makes sense

**Comment by aljoscha** [comment]
Thanks! I think I can merge once Travis is green. 👌

**Comment by ifndef-SleePy** [comment]
Travis failed, there is am importing error in python module. I don't think it is caused by this PR. I will try to rebase master and trigger travis checking again.

**Comment by ifndef-SleePy** [comment]
Python module is fine but connectors module failed this time on Travis checking. It seems a travis relevant issue. I have filed a JIRA issue for this, https://issues.apache.org/jira/browse/FLINK-12866
Will trigger checking again through close and reopening, hope it works.

**Comment by aljoscha** [comment]
I merged this. Thanks for working on this with me! 👍
